src/runtime/program/precompile_programs/secp256k1.zig

```diff
@@ -43,8 +49,595 @@ pub fn verify(
     current_instruction_data: []const u8,
     all_instruction_datas: []const []const u8,
 ) PrecompileProgramError!void {
-    _ = current_instruction_data;
-    _ = all_instruction_datas;
+    const data = current_instruction_data;
+
+    if (data.len < SECP256K1_DATA_START) {
+        if (data.len == 1 and data[0] == 0) return; // success
+        return error.InvalidInstructionDataSize;
+    }
+
+    const n_signatures = data[0];
+    if (n_signatures == 0 and data.len > 1) return error.InvalidInstructionDataSize;
+
+    const expected_data_size = SECP256K1_SIGNATURE_OFFSETS_START +
+        n_signatures * SECP256K1_SIGNATURE_OFFSETS_SERIALIZED_SIZE;
+    if (data.len < expected_data_size) return error.InvalidInstructionDataSize;
+
+    for (0..n_signatures) |i| {
+        const offset = SECP256K1_SIGNATURE_OFFSETS_START +
+            i * SECP256K1_SIGNATURE_OFFSETS_SERIALIZED_SIZE;
+        const sig_offsets: *align(1) const Secp256k1SignatureOffsets = @alignCast(
+            @ptrCast(data.ptr + offset),
+        );
+
+        // This case isn't useful, but agave has it.
+        if (sig_offsets.signature_instruction_idx >= all_instruction_datas.len) {
+            return error.InvalidInstructionDataSize;
+        }
+
+        const signature_slice = try getInstructionData(
+            SECP256K1_SIGNATURE_SERIALIZED_SIZE + 1, // + recovery_id
+            all_instruction_datas,
+            sig_offsets.signature_instruction_idx,
+            sig_offsets.signature_offset,
+        );
+
+        const recovery_id: u2 = blk: {
+            const rec_id = signature_slice[SECP256K1_SIGNATURE_SERIALIZED_SIZE];
+            if (rec_id > 3) return error.InvalidRecoveryId;
+            break :blk @intCast(rec_id);
+        };
+        // https://docs.rs/libsecp256k1/0.6.0/src/libsecp256k1/lib.rs.html#674-680
+
+        const signature: *const Ecdsa.Signature = @ptrCast(
+            signature_slice[0..SECP256K1_SIGNATURE_SERIALIZED_SIZE],
+        );
+
+        const eth_address = try getInstructionData(
+            SECP256K1_PUBKEY_SERIALIZED_SIZE,
```

Comment:
`SECP256K1_ETH_ADDRESS_SERIALIZED_SIZE` -- note: kinda confusing 'eth pubkey' is just a few bytes of the hashed value of an actual pubkey -- maybe we stick to eth_address everywhere (ie, `constructEthPubkey` -> `constructEthAddress`)